addons/project/tests/test_multicompany.py

```diff
@@ -220,7 +220,126 @@ def test_generate_analytic_account(self):
                 self.project_company_a._create_analytic_account()
 
-                self.assertEqual(self.project_company_a.company_id, self.project_company_a.analytic_account_id.company_id, "The analytic account created from a project should be in the same company")
+                self.assertEqual(self.project_company_a.company_id, self.project_company_a.analytic_account_id.company_id, "The analytic account created from a project should be in the same company.")
+
+        project_no_company = self.Project.create({'name': 'Project no company'})
+        plans = self.env['account.analytic.plan'].sudo().search([('company_id', '=', False)])
+        #ensures that all the existing plan have a company_id
+        for plan in plans:
+            plan.company_id = self.env.company
+        project_no_company._create_analytic_account()
+        self.assertFalse(project_no_company.analytic_account_id.company_id, "The analytic account created from a project without company_id should have its company_id field set to False.")
+        self.assertFalse(project_no_company.analytic_account_id.plan_id.company_id, "The analytic plan created from a project without company_id should have its company_id field set to False.")
+        plan = self.env['account.analytic.plan'].sudo().search([('company_id', '=', False)])
+        self.assertEqual(1, len(plan), "Only one analytic plan should have been created.")
+
+        project_no_company_2 = self.Project.create({'name': 'Project no company 2'})
+        project_no_company_2._create_analytic_account()
+        self.assertNotEqual(project_no_company_2.analytic_account_id, project_no_company.analytic_account_id, "The analytic account created should be different from the account created for the 1st project.")
+        plan = self.env['account.analytic.plan'].sudo().search([('company_id', '=', False)])
+        self.assertEqual(1, len(plan), "No new analytic should have been created.")
+
+    def test_analytic_account_company_consistency(self):
+        """
+            This test ensures that the following invariant is kept:
+            If the company of an analytic account is set, all of its project must have the same company.
+            If the company of an analytic account is not set, its project can either have a company set, or none.
+        """
+        project_no_company = self.Project.create({'name': 'Project no company'})
+        project_no_company._create_analytic_account()
+        account_no_company = project_no_company.analytic_account_id
+        self.project_company_a._create_analytic_account()
+        account_a = self.project_company_a.analytic_account_id
+
+        # Set the account of the project to a new account without company_id
+        self.project_company_a.analytic_account_id = account_no_company
+        self.assertEqual(self.project_company_a.analytic_account_id, account_no_company, "The new account should be set on the project.")
```

Comment:
Actually, its more like a security mesure. I've added the _inverse_company_id for projects. And in this method, the company of the project is (in some case) revert back. This assert is there to ensure the company_id of the project was not reverted back. Though, since in cases the revert is done a user error is raised, maybe we could indeed just remove the assert and let the test fails because of the error raised. But i'd rather keep it that way in case the raise error is removed in future updated.